airflow/settings.py

```diff
@@ -27,7 +27,7 @@
 from typing import TYPE_CHECKING, Any, Callable
 
 import pluggy
-from sqlalchemy import create_engine, exc, text
+from sqlalchemy import __version__ as sqlalchemy_version, create_engine, exc, text
```

Comment:
We don't have a discussion yet about whether or not we should use `__version__` or `importlib.metadata`.
@potiuk any thoughts? I'm personally have no objections to use it for now, because SQLachemy maintainers do not announced about deprecate/remove `__version__`

Reply:
I originally did this (in a new util file dedicated to dependency version checks):
```python
from packaging.version import Version, parse as parse_version
from sqlalchemy import __version__ as sqlalchemy_version


def is_sqlalchemy_v1() -> bool:
    """Check if SQLAlchemy v1 is installed."""
    installed: Version = parse_version(sqlalchemy_version)
    return installed.release[0] == 1
```

But then saw the `PENDULUM3` example and amended my code. I don't changing it to anything else.

Reply:
No I mean a bit different, e.g. get from distribution package, but we do not discuss preferable way

```python
import sys
if sys.version_info >= (3, 9):
    from importlib.metadata import version
else:
    from importlib_metadata import version

SQL_ALCHEMY_V1 = version("sqlalchemy").startswith("1")
```


> But then saw the PENDULUM3 example and amended my code.

Pendulum3 I guess exists only in tests

Reply:
1. [Not only in tests](https://github.com/apache/airflow/blob/016892b7d6bc654ac3ff7e4f8f9f7231f835bed7/airflow/utils/timezone.py#L32).
2. Importlib metadata was [added in py3.8,](https://docs.python.org/3/library/importlib.metadata.html) which is the minimum aupported in airflow anyway, so no need to do any fancy checking.

Reply:
> [Not only in tests](https://github.com/apache/airflow/blob/016892b7d6bc654ac3ff7e4f8f9f7231f835bed7/airflow/utils/timezone.py#L32).

Oh... yeah... that reminds me that we decide to not to use try except. But most possible pendulum 2 would be dropped into the Airflow 2.9 or 2.10

> Importlib metadata was [added in py3.8,](https://docs.python.org/3/library/importlib.metadata.html) which is the minimum supported in airflow anyway, so no need to do any fancy checking.

Importlib mandatory since 3.10, in cpython, I guess, most of the things implemented in 3.9. And something more, airflow read entire environment metadata during the providers loading, so this not a big deal even cache all packages installed into the environment. 

Anyway this one just a suggestion, no objections at this point. It could be changed in the future, we still not fully support SA2.0, so we have plenty a time. The point here, that some of package stop provide `__version__` (e.g. Flask and everything related to it) or newer provide it.

Reply:
I ended up using try/except after trying, without success, to mock an engine (using MagicMock, and sqlalchemy.engine.mock). I couldn't get past the event registration which throws the TypeError I ended up catching. It was either that or get rid if the test entirely. If you (or anyone else) has an idea how to write this test w/o try/except - I'll be happy to hear.

BTW, what's wrong with try/except? I thought EAFP was the preferred way to write python code.

Reply:
> I thought EAFP was the preferred way to write python code.

Nothing wrong, did I tell that it wrong? In general everything depends on the situation, depend on the cases. There was more hacky case rather than check the version

Reply:
In this case it is fine. 

BTW, could you rebase your branch? It should fix failing tests.